django/db/models/sql/query.py

```diff
@@ -2108,3 +2108,435 @@ def update_join_types(self, query):
+        self.check_related_objects(field, value, opts)
+
+        targets = self.trim_joins(sources, path)
+
+        if field.is_relation:
+            # No support for transforms for relational fields
+            num_lookups = len(lookups)
+            if num_lookups > 1:
+                raise FieldError('Related Field got invalid lookup: {}'.format(lookups[0]))
+            assert num_lookups > 0  # Likely a bug in Django if this fails.
+            lookup_class = field.get_lookup(lookups[0])
+            if lookup_class is None:
+                raise FieldError('Related Field got invalid lookup: {}'.format(lookups[0]))
+            assert len(targets) == 1
+            lhs = SimpleCol(targets[0], field)
+            condition = lookup_class(lhs, value)
+            lookup_type = lookup_class.lookup_name
+        else:
+            col = SimpleCol(targets[0], field)
+            condition = self.build_lookup(lookups, col, value)
+            lookup_type = condition.lookup_name
+
+        clause.add(condition, AND)
+
+        if current_negated and lookup_type != 'isnull' and self.is_nullable(targets[0]):
+            # The condition added here will be SQL like this:
+            # NOT (col IS NOT NULL), where the first NOT is added in
+            # upper layers of code. The reason for addition is that if col
+            # is null, then col != someval will result in SQL "unknown"
+            # which isn't the same as in Python. The Python None handling
+            # is wanted, and it can be gotten by
+            # (col IS NULL OR col != someval)
+            #   <=>
+            # NOT (col IS NOT NULL AND col = someval).
+            lookup_class = targets[0].get_lookup('isnull')
+            clause.add(lookup_class(SimpleCol(targets[0], field), False), AND)
+        return clause
+
+    def build_where(self, q_object):
+        """
+        A preprocessor for the internal _add_q(). Responsible for doing final
+        join promotion.
+        """
+        return self._add_q(q_object)
+
+    def _add_q(self, q_object, branch_negated=False, current_negated=False):
+        """Add a Q-object to the current filter."""
+        connector = q_object.connector
+        current_negated = current_negated ^ q_object.negated
+        branch_negated = branch_negated or q_object.negated
+        target_clause = WhereNode(connector=connector, negated=q_object.negated)
+        for child in q_object.children:
+            if isinstance(child, Node):
+                child_clause = self._add_q(child, branch_negated, current_negated)
+            else:
+                child_clause = self.build_filter(
+                    child, branch_negated=branch_negated,
+                    current_negated=current_negated,
+                )
+            if child_clause:
+                target_clause.add(child_clause, connector)
+        return target_clause
+
+    def names_to_path(self, names, opts, allow_many=True, fail_on_missing=False):
```

Comment:
I've not put any significant effort into understanding this method yet, so it may well be possible to simplify.